### DISCUSSION TITLE
Deep Knowledge Tracing (Piech et al, 2015) 리뷰

### DISCUSSION THREAD

**Discussion by sooyounlee** [bcaitech1/p4-dkt-DKTJHGSD#61] [Show and tell]
## Abstract

- Computer를 이용한 교육에서 Knowledge Tracing은 잘 정립되어 있는 problem이며, 내재적으로 한계점들이 존재했음
- 이 paper에서는 Recurrent Neural Network(RNN)을 이용하여 student learning을 모델링함
- RNN이 기존 방법론에 비해 가지는 이점은 다음과 같음:
    - 분야에 대한 지식을 직접적으로 encoding해줄 필요가 없음
    - 더 complex한 학생의 지식을 포착할 수 있음
    - RNN을 이용한 knowledge tracing dataset에 대한 예측 성능이 이전에 비해 훨씬 향상되었음
- 학습 모델은 curriculum design이나 학생의 과업들에 대한 직관적인 해석이나 발견을 가능하게 함

## 1. Introduction

- 컴퓨터의 발전에 따라 Khan academy, Coursera, EdX와 같이 수준 높은 교육에 대한 접근성을 높여주는 서비스들이 등장함
- Knowledge tracing은 학생의 지식 수준을 잘 모델링하여 학생이 미래에 해당 분야에 대한 지식을 얼마나 잘 보일지(문제를 잘 풀지)를 정확히 예측하고자 하는 task임
- 이 task를 잘 수행하게 되면:
    - 학생 각자의 필요에 따라 resource를 제안해주는 것이 가능해짐
    - 너무 쉽거나 너무 어렵다고 판단되는 학습 내용들은 뛰어넘거나 뒷부분으로 미루는 것이 가능해짐
- 이미 인간의 판단으로 curate된 tutoring으로부터 좋은 성과가 나온다는 것이 알려짐: 일대일 과외는 평균적인 학생의 약 2 표준편차만큼 증가시켜줌 (50% → 97%?)
![image](https://user-images.githubusercontent.com/21062594/120012100-075ab580-c01a-11eb-9386-041b42e43498.png)

- 머신러닝을 이용하여 어느 정도 knowledge tracing에서 성과를 내고 있었음
- 머신러닝을 이용한 KT의 한계점:
    - 학습이라는 것이 본질적으로 인간의 뇌와 인간의 지식의 복잡성에 기반하고 있음

        → 그러므로 rich model(복잡한 모델?)을 사용하는 것이 좀 더 적절해 보임

    - 하지만 기존 대부분의 방법론은 first-order Markov model을 기반으로 하고 있음

- 이 논문에서는 Deep Learning을 이용한 Deep Knowledge Tracing(DKT)을 선보임
    - RNN을 사용
    - latent knowledge state, temporal dynamics을 인공 '신경망'을 사용하여 표현하였으며, 직접 rule을 만들기보다 데이터를 학습시켜 학생 지식에 대한 잠재 변수들이 표현될 수 있음
- 이 논문의 주요한 기여는 다음과 같음:
    - KT 분야에서 RNN의 첫 적용
    - KT 벤치마크에서 최고 성능 대비 25% AUC 상승
    - 전문가가 참여하여 annotation할 필요가 없음을 보여줌
    - 연습의 영향을 발견하였으며 연습을 잘 할 수 있는 진보된 커리큘럼들을 만들어냄
- Knowledge tracing?
    - 8학년 여학생이 수학을 배우는 것에 대한 지식 변화를 추적한 것을 보여주는 그래프
  
![image](https://user-images.githubusercontent.com/21062594/120012172-20636680-c01a-11eb-99d2-d53b51a58d02.png)

- 먼저 Square roots와 x-intercept 관련된 한 문제를 맞게 푼 이후, x-intercept, y-intercept, graphing과 관련된 문제들을 풀었음
- 매번 연습 문제에 답할 때마다 다음 문제를 맞출 것인지를 예측할 수 있음
- 기존에는 전문가가 각 문제에 대해 주제를 직접 tag해주었으나, RNN을 이용하면 그러한 tag가 불필요함 - 모델이 자율적으로 학습 내용의 구조를 학습할 수 있음

## 2. Related Works

- 다양한 분야 - 교육, 심리학, 신경과학, 인지과학 등 - 과 연관되어 있음
- 사회과학에서는 학습이 거시적인 요소들에 영향을 받는다고 함: 정서, 동기(motivation), 심지어 정체성 등
    - 현재 해당 분야에서의 과제들은 미시적인 요소들에 대한 영향과의 관련성
- 인지과학에서는 학습과 관련되어 두 가지 주제가 특히 관련이 높음: 인간의 마음과 학습 과정은 recursive (involving doing or saying the same thing several times in order to produce a particular result or effect) 하며, 유추에 기반함
- 이러한 논의들에 기반해서 사교육 업계에서는 모든 인지적 과정을 capture하지는 못하더라도 유용한 모델을 개발하는 것이 우선 과제였음

### 2.3 Recurrent Neural Network

- 기존 사용되던 Markov 모델도 dynamic했으나, 이에 비해 RNN은 지식 잠재 상태에 대해 더 고차원이고, 연속적으로 나타낼 수 있었음.
- RNN의 장점은 input의 정보를 시간이 지난 후에도 사용할 수 있다는 것임 - 특히 LSTM 네트워크들이 이런 장점을 가지고 있음

## 3. Deep Knowledge Tracing

- 인간 학습은 다양한 요소들에 영향을 받음 - 학습 자료, 학습 내용, 학습 시간, 학습에 연관된 사람들: 계량화하기가 어려움
- 이 논문에서는 두 가지 RNN 모델을 사용
    - Vanilla RNN with sigmoid
    - LSTM

### 3.1 Model

- RNN은 input sequence를 통해 output sequence를 예측하는 데 hidden state들의 sequence를 계산하여 사용하는데, 이러한 hidden state들을 과거 학생 학습에 대한 관찰 중 미래 예측에 유용한 정보들을 잘 embedding한 것이라고 볼 수 있음
![image](https://user-images.githubusercontent.com/21062594/120012211-2d805580-c01a-11eb-96f6-19262e10d9bc.png)
![image](https://user-images.githubusercontent.com/21062594/120012236-34a76380-c01a-11eb-8378-8023b3ac86b9.png)

- LSTM은 RNN의 더 복잡한 변형으로, 종종 더 강력하다고 평가되고 있음
    - RNN에 비해 과거의 정보를 더 오래 지니고 있음

### 3.2 Input and Output time series

- RNN을 이용하기 위해 interaction들을 정해진 길이의 벡터의 sequence로 바꿀 필요가 있음. 이 논문에서는 두 가지 방법을 사용함:
    1. 작은 데이터셋에서는 해당 질문을 맞췄는지 틀렸는지만에 대해 one-hot encoding을 함. question과 answer를 따로 벡터화해서 넣는 것이 오히려 성능을 하락시킴.
    2. feature가 많은 경우, one-hot encoding 때문에 벡터 크기가 너무 커질 수 있음. 그러므로 대신 각각 랜덤하게 vector를 지정해줄 수 있음.
![image](https://user-images.githubusercontent.com/21062594/120012414-6f110080-c01a-11eb-91a7-e62254cb8d02.png)

- output vector y는 문제의 수와 같은 길이의 벡터로, 각 숫자는 학생이 해당 문제를 맞출지에 대한 예측 확률을 나타냄

### 3.3 Optimization

- Loss: binary cross entropy
    - Loss for a single student is:
![image](https://user-images.githubusercontent.com/21062594/120012429-759f7800-c01a-11eb-8fe5-58e97b396bf7.png)
- SGD
- overfitting을 피하기 위해 y 계산 시 ht에 dropout 적용 (ht+1 계산 시에는 적용하지 않음)
- gradient clipping
- hidden dimension: 200
- mini-batch: 100

## 5. Datasets

- simulated data

5-fold

- Khan academy data
- Assistments benchmark dataset
![image](https://user-images.githubusercontent.com/21062594/120012462-8223d080-c01a-11eb-833c-65b1b700b3af.png)

## 6. Results

- 기존 방법론들보다 나은 결과
![image](https://user-images.githubusercontent.com/21062594/120012532-9667cd80-c01a-11eb-9129-8c37c13676d6.png)

- 학생들이 문제를 푼 순서를 통해 문제 간 연관성을 파악할 수 있었음 (전문가의 annotation 없이!)
![image](https://user-images.githubusercontent.com/21062594/120012579-a384bc80-c01a-11eb-930f-cb339ffb4dfb.png)

- weight clustering은 크게 쓸모있지 않았음
![image](https://user-images.githubusercontent.com/21062594/120012610-af707e80-c01a-11eb-91fc-86aaba67469d.png)
